src/test/librbd/test_Groups.cc

```diff
@@ -480,3 +480,241 @@ TEST_F(TestGroup, add_snapshotPP)
+      ASSERT_STREQ(image_name1, gp_snaps[i].image_snaps[0].image_name);
+      ASSERT_EQ(rados_ioctx_get_id(ioctx), gp_snaps[i].image_snaps[0].pool_id);
+    } else if (!strcmp(gp_snaps[i].name, gp_snap_name2)) {
+      ASSERT_EQ(1U, gp_snaps[i].image_snaps_count);
+      ASSERT_STREQ(image_name2, gp_snaps[i].image_snaps[0].image_name);
+      ASSERT_EQ(rados_ioctx_get_id(ioctx2), gp_snaps[i].image_snaps[0].pool_id);
+    } else {
+      FAIL() << "Unexpected group snap: " << gp_snaps[i].name;
+    }
+  }
+  ASSERT_EQ(0, rbd_group_snap_remove(ioctx, gp_name, gp_snap_name1));
+  ASSERT_EQ(0, rbd_group_snap_remove(ioctx, gp_name, gp_snap_name2));
+  ASSERT_EQ(0, rbd_group_remove(ioctx, gp_name));
+}
+
+TEST_F(TestGroup, snap_list2PP)
+{
+  REQUIRE_FORMAT_V2();
+
+  std::string pool_name2 = get_temp_pool_name("test-librbd-");
+  ASSERT_EQ(0, _rados.pool_create(pool_name2.c_str()));
+
+  librados::IoCtx ioctx2;
+  ASSERT_EQ(0, _rados.ioctx_create(pool_name2.c_str(), ioctx2));
+
+  const char *image_name1="snap_list2PP_image1";
+  ASSERT_EQ(0, create_image_pp(m_rbd, m_ioctx, image_name1, m_image_size));
+
+  const char *image_name2="snap_list2PP_image2";
+  ASSERT_EQ(0, create_image_pp(m_rbd, ioctx2, image_name2, m_image_size));
+
+  const char *gp_name= "snap_group";
+  ASSERT_EQ(0, m_rbd.group_create(m_ioctx, gp_name));
+
+  std::vector<librbd::group_snap_info2_t> gp_snaps;
+  ASSERT_EQ(0, m_rbd.group_snap_list2(m_ioctx, gp_name, &gp_snaps));
+  ASSERT_EQ(0U, gp_snaps.size());
+
+  const char *gp_snap_name0 = "snap_snapshot0";
+  ASSERT_EQ(0, m_rbd.group_snap_create(m_ioctx, gp_name, gp_snap_name0));
+
+  const char *gp_snap_name1 = "snap_snapshot1";
+  ASSERT_EQ(0, m_rbd.group_image_add(m_ioctx, gp_name, m_ioctx,
+                                     image_name1));
+  ASSERT_EQ(0, m_rbd.group_snap_create(m_ioctx, gp_name, gp_snap_name1));
+
+  const char *gp_snap_name2 = "snap_snapshot2";
+  ASSERT_EQ(0, m_rbd.group_image_remove(m_ioctx, gp_name,
+                                        m_ioctx, image_name1));
+  ASSERT_EQ(0, m_rbd.group_image_add(m_ioctx, gp_name, ioctx2,
+                                     image_name2));
+  ASSERT_EQ(0, m_rbd.group_snap_create(m_ioctx, gp_name, gp_snap_name2));
+
+  ASSERT_EQ(0, m_rbd.group_snap_list2(m_ioctx, gp_name, &gp_snaps));
+  ASSERT_EQ(3U, gp_snaps.size());
+
+  for (const auto &i : gp_snaps) {
+    if (i.name == gp_snap_name0) {
+      ASSERT_EQ(0U, i.image_snaps.size());
+    } else if (i.name == gp_snap_name1) {
+      ASSERT_EQ(1U, i.image_snaps.size());
+      ASSERT_EQ(image_name1, i.image_snaps[0].image_name);
+      ASSERT_EQ(m_ioctx.get_id(), i.image_snaps[0].pool_id);
+    } else if(i.name == gp_snap_name2) {
+      ASSERT_EQ(1U, i.image_snaps.size());
+      ASSERT_EQ(image_name2, i.image_snaps[0].image_name);
+      ASSERT_EQ(ioctx2.get_id(), i.image_snaps[0].pool_id);
+    } else {
+      FAIL() << "Unexpected group snap: " << i.name;
+    }
+  }
+
```

Comment:
It would be nice to list the snaps again after removing one or more snaps to make sure they have been removed.